pdns/opensslsigners.cc

```diff
@@ -406,42 +677,100 @@ int OpenSSLRSADNSCryptoKeyEngine::hashSizeToKind(const size_t hashSize)
 
 std::string OpenSSLRSADNSCryptoKeyEngine::sign(const std::string& message) const
 {
+  std::string signature;
+
+#if OPENSSL_VERSION_MAJOR >= 3
+  auto ctx = MessageDigestContext(EVP_MD_CTX_new(), EVP_MD_CTX_free);
+  if (ctx == nullptr) {
+    throw pdns::OpenSSL::error(getName(), "Could not create context for signing");
+  }
+
+  if (EVP_DigestSignInit(ctx.get(), nullptr, hasher(), nullptr, d_key.get()) == 0) {
+    throw pdns::OpenSSL::error(getName(), "Could not initialize context for signing");
+  }
+
+  std::size_t signatureLen = 0;
+  // NOLINTNEXTLINE(*-cast): Using OpenSSL C APIs.
+  auto* messageCStr = const_cast<unsigned char*>(reinterpret_cast<const unsigned char*>(message.c_str()));
+  if (EVP_DigestSign(ctx.get(), nullptr, &signatureLen, messageCStr, message.size()) == 0) {
+    throw pdns::OpenSSL::error(getName(), "Could not get message signature length");
+  }
+
+  signature.resize(signatureLen);
+
+  // NOLINTNEXTLINE(*-cast): Using OpenSSL C APIs.
+  auto* signatureCStr = const_cast<unsigned char*>(reinterpret_cast<const unsigned char*>(signature.c_str()));
+  if (EVP_DigestSign(ctx.get(), signatureCStr, &signatureLen, messageCStr, message.size()) == 0) {
+    throw pdns::OpenSSL::error(getName(), "Could not sign message");
+  }
+#else
+  unsigned int signatureLen = 0;
   string l_hash = this->hash(message);
   int hashKind = hashSizeToKind(l_hash.size());
-  std::string signature;
   signature.resize(RSA_size(d_key.get()));
-  unsigned int signatureLen = 0;
 
+  // NOLINTNEXTLINE(*-cast): Using OpenSSL C APIs.
   int res = RSA_sign(hashKind, reinterpret_cast<unsigned char*>(&l_hash.at(0)), l_hash.length(), reinterpret_cast<unsigned char*>(&signature.at(0)), &signatureLen, d_key.get());
   if (res != 1) {
     throw runtime_error(getName() + " failed to generate signature");
   }
 
   signature.resize(signatureLen);
+#endif
+
   return signature;
 }
 
-
 bool OpenSSLRSADNSCryptoKeyEngine::verify(const std::string& message, const std::string& signature) const
 {
+#if OPENSSL_VERSION_MAJOR >= 3
+  auto ctx = MessageDigestContext(EVP_MD_CTX_new(), EVP_MD_CTX_free);
+  if (ctx == nullptr) {
+    throw pdns::OpenSSL::error(getName(), "Failed to create context for verifying signature");
+  }
+
+  if (EVP_DigestVerifyInit(ctx.get(), nullptr, hasher(), nullptr, d_key.get()) == 0) {
+    throw pdns::OpenSSL::error(getName(), "Failed to initialize context for verifying signature");
+  }
+
+  // NOLINTNEXTLINE(*-cast): Using OpenSSL C APIs.
+  const int ret = EVP_DigestVerify(ctx.get(), reinterpret_cast<const unsigned char*>(signature.data()), signature.size(), reinterpret_cast<const unsigned char*>(message.data()), message.size());
+  if (ret == 0) {
+    throw pdns::OpenSSL::error(getName(), "Signature did not verify successfully: Either the given message does not match the original message, or the signature has an invalid form");
```

Comment:
Should we really throw on failure to validate the signature? Would returning `false` not be enough?